### PR TITLE
PLAT-1730 - Refactor Guzzle Cache Plugin Out

### PR DESCRIPTION
**Pull request by markwallsgrove**
## problem
The caching mechanism is implemented by using a `CachePlugin` within Guzzle which operates at the HTTP transport layer. When a exception occurs within the plugin it propagates through Guzzle causing the request to fail.
## fix
To be able to control the exceptions which are propogated from the caching mechanism Guzzle and the cache need to be separated. This means that we will be able to control what happens if the cache fails. 

The requirement is to be able to fallback to Persona for the public certificate and token verification if the cache is not accessible. 

fixes https://github.com/talis/platform/issues/1730

### CODE REVIEW

File: src/Talis/Persona/Client/Base.php [hunk 1]
```diff
@@ -289,7 +260,8 @@ protected function createRequest($url, array $opts)
             $httpConfig['headers']['Content-Type'] = 'application/x-www-form-urlencoded';
         }
 
-        $client = $this->getHTTPClient($opts['skipRevalidation']);
+        // TODO: turn off all caching
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Base.php]: do the plugins need configuring? I believe the cache plugins are defaulted, but the no-cache header is set. Review the behavior.

File: src/Talis/Persona/Client/Base.php [hunk 2]
```diff
@@ -33,9 +35,6 @@ abstract class Base
      */
     private $logger;
 
-     /** @var \Talis\Persona\Client\HttpClientFactory */
-    private $httpClientFactory;
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Base.php]: the httpClientFactory is no longer needed, so it has  been removed

File: src/Talis/Persona/Client/Base.php [hunk 3]
```diff
@@ -177,53 +162,6 @@ protected function getLogger()
...
-    protected function getHTTPClient($skipRevalidation = false)
-    {
-        return $this->httpClientFactory->create($skipRevalidation);
-    }
-
-    /**
-     * Retrieve the Persona client version
-     * @return string Persona client version
-     */
-    protected function getClientVersion()
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Base.php]: moved to ClientVersionCache

File: src/Talis/Persona/Client/Base.php [hunk 4]
```diff
@@ -253,13 +201,13 @@ protected function createRequest($url, array $opts)
 
         $opts = array_merge(
             [
-                'headers' => [],
+                'headers' => [
+                    'Cache-Control' => 'max-age=0, no-cache',
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Base.php]: all calls state that caching shouldn't be used at a http level

File: src/Talis/Persona/Client/OAuthClients.php [hunk 5]
```diff
@@ -9,12 +9,11 @@ class OAuthClients extends Base
      *
      * @param string $clientId Persona client id
      * @param string $token Persona client token
-     * @param integer $cacheTTL time to live in seconds value for cached request
      * @return array response from Persona
      * @throws \InvalidArgumentException Invalid arguments
      * @throws \Exception Persona communication issues
      */
-    public function getOAuthClient($clientId, $token, $cacheTTL = 300)
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/OAuthClients.php]: cacheTTL was related to the HTTP cache expiry time

File: src/Talis/Persona/Client/Tokens.php [hunk 6]
```diff
@@ -61,14 +63,15 @@ public function validateToken(array $params = [])
...
      * @throws ScopesNotDefinedException If the JWT token doesn't include the user's scopes
      * @throws Exception If not able to communicate with Persona to retrieve the public certificate
      */
-    protected function validateTokenUsingJWT($token, array $scopes = null, $cacheTTL = 300)
+    protected function validateTokenUsingJWT($token, array $scopes = null)
     {
-        $publicCert = $this->retrieveJWTCertificate($cacheTTL);
+        $publicCert = $this->retrieveJWTCertificate();
+
+        // TODO: check certificate?
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Tokens.php]: TODO

File: src/Talis/Persona/Client/Tokens.php [hunk 7]
```diff
@@ -166,7 +175,7 @@ protected function validateTokenUsingPersona($token, array $scopes = [])
         $success = $this->personaCheckTokenIsValid($token, $scopes);
         $this->getStatsD()->endTiming('validateToken.rest.get');
 
-        if ($success === true) {
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Tokens.php]: bug: `personaCheckTokenIsValid` returns `ValidationResults` enum

File: src/Talis/Persona/Client/Tokens.php [hunk 8]
```diff
@@ -282,14 +290,7 @@ protected function checkConfig(array $config)
     protected function makePersonaHttpRequest($url)
     {
         try {
-            $body = $this->performRequest(
-                $url,
-                [
-                    'headers' => [
-                        'Cache-Control' => 'max-age=0, no-cache',
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Tokens.php]: moved within performRequest

File: test/unit/Persona/TokensTest.php [hunk 9]
```diff
@@ -956,45 +958,4 @@ public function testListScopesScopeCountInvalidDomain()
         $this->setExpectedException(InvalidTokenException::class);
         $mockClient->listScopes($accessToken);
     }
-
-   public function testRetrieveJWTCertificateSkipsRevalidation()
```
Comment (markwallsgrove) [inline test/unit/Persona/TokensTest.php]: was testing skipping http cache revalidation

File: src/Talis/Persona/Client/Tokens.php [hunk 10]
```diff
@@ -60,15 +69,17 @@ public function validateToken(array $params = [])
...
-     * @param array|null $scopes specify this if you wish to validate a scoped token
-     * @param integer $cacheTTL time to live value in seconds for the certificate to stay within cache
-     * @return integer ValidationResults enum
      * @throws ScopesNotDefinedException If the JWT token doesn't include the user's scopes
-     * @throws Exception If not able to communicate with Persona to retrieve the public certificate
+     * @throws CommunicationException Cannot communicate with Persona
+     * @throws \Exception If not able to communicate with Persona to retrieve the public certificate
      */
-    protected function validateTokenUsingJWT($token, array $scopes = null, $cacheTTL = 300)
+    protected function validateTokenUsingJWT($token, array $scopes = null)
```
Comment (ShaunParsons) [inline src/Talis/Persona/Client/Tokens.php]: Why are we removing the ability to check the TTL?

Comment (markwallsgrove) [reply src/Talis/Persona/Client/Tokens.php]: the `$cacheTTL` attribute was related to the HTTP caching layer

**Pull request merged**